### PR TITLE
`variable_scope` use `auxiliary_name_scope` to control whether to create new name scope

### PR DESCRIPTION
**Pull request by facaiy**
Now `variable_scope` always create a new name scope (side-effect?) when invoked. The behavior will result in name scope collision mentioned in #13429 . Hence we proposed to add a parameter `auxiliary_name_scope` to control whether to create new name scope or not.

If accepted, we can reenter `variable_scope` and its original name scope without side-effect:
```python
with tf.variable_scope(s, auxiliary_name_scope=False) as ss:
     with tf.name_scope(ss.original_name_scope) as n:
          # do something
```
### How to test

+ [x] add unit tests
+ [ ] pass all tests.

### CODE REVIEW

**Comment by facaiy** [comment]
Since the PR requires API review, cc @lukaszkaiser @martinwicke

**Review by lukaszkaiser — APPROVED** [review]
These changes look good to me, thanks!

**Comment by lukaszkaiser** [comment]
The code is good and I think this addresses a bunch of needs in probably the simplest possible way, thanks! Still, it's an API change, need to wait for Martin (and maybe others) to decide.

File: tensorflow/python/ops/variable_scope.py [hunk 1]
```diff
@@ -1832,6 +1839,19 @@ def __enter__(self):
       if self._current_name_scope is not None:
         self._current_name_scope.__enter__()
       return self._cached_pure_variable_scope.__enter__()
+    if self._auxiliary_name_scope:
+      # Create a new name scope later
+      current_name_scope = None
+    else:
+      # Reuse the current name scope
+      name_scope = ops.get_name_scope()
```
Comment (facaiy) [inline tensorflow/python/ops/variable_scope.py]: I'm afraid the `ops.get_name_scope()` might be a performance bottleneck here.

**Comment by facaiy** [comment]
I revise `auxiliary_name_scope` to support name scope argument, and summary as below:
+ `auxiliary_name_scope=True`, create new name scope as before by default.
+ `auxiliary_name_scope=None` or `False`, don't create any name scope.
+ `auxiliary_name_scope=some_name_scope`, reuse it if `some_name_scope`  is `name_scope`.

Now we can write one line code to reuse both variable_scope and its name_scope:

```python
with tf.variable_scope(s, auxiliary_name_scope=s.original_name_scope) as ss:
    # do something
```
or
```python
with tf.variable_scope(s, auxiliary_name_scope=None) as ss:
     with tf.name_scope(ss.original_name_scope) as n:
          # do something
```

**Comment by facaiy** [comment]
@lukaszkaiser Might you take a look again? Thanks.

**Review by lukaszkaiser — APPROVED** [review]
I liked it better without the newly added complexity, but it's still fine.

**Comment by facaiy** [comment]
Thanks. I agree with you that the commits later introduce too much complexity. I'd like to revert it to the first version which only supports boolean. And if name scope is requested in the future, we could add the function easily later.

**Comment by lukaszkaiser** [comment]
Yes, that'd be great.

**Comment by facaiy** [comment]
I have reverted codes to boolean version:
+ `auxiliary_name_scope=True`, create new name scope as before by default.
+ `auxiliary_name_scope=False`, don't create any name scope.

File: tensorflow/python/ops/variable_scope.py [hunk 2]
```diff
@@ -1832,6 +1839,21 @@ def __enter__(self):
       if self._current_name_scope is not None:
         self._current_name_scope.__enter__()
       return self._cached_pure_variable_scope.__enter__()
+
+    if self._auxiliary_name_scope:
+      # Create a new name scope later
+      current_name_scope = None
+    else:
+      # Reenter the current name scope
+      name_scope = ops.get_name_scope()
```
Comment (facaiy) [inline tensorflow/python/ops/variable_scope.py]: I just worry that `get_name_scope` here might be performance bottleneck.

File: tensorflow/python/kernel_tests/variable_scope_test.py [hunk 3]
```diff
@@ -732,6 +732,115 @@ def testVarOpScopeNestedOuterScope(self):
           with ops.name_scope("scope2") as sc2:
             self.assertEqual(sc2, "outer_1/default/scope2/")
 
+  def testBasicWhenAuxiliaryNameScopeIsFalse(self):
+    with self.test_session():
+      with variable_scope.variable_scope("scope",
+                                         auxiliary_name_scope=False) as scope:
+        self.assertEqual(scope.original_name_scope, "")
```
Comment (facaiy) [inline tensorflow/python/kernel_tests/variable_scope_test.py]: I think the `original_name_scope` is reasonable (but counterintuitive?) behavior. We almost never create `variable_scope` like this case. I'm afraid that user may misuse the argument and get lost in name scope.

Comment (facaiy) [reply tensorflow/python/kernel_tests/variable_scope_test.py]: Perhaps it's better to restrict `auxiliary_name_scope` for reentering case.  Namely, `auxiliary_name_scope` only takes effect when VariableScope is passed.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
Github failure?

```bash
Build was aborted
Aborted by unknown
Unable to get pull request builder trigger!!
```

**Comment by facaiy** [comment]
unrelated failure?
`Invocation: Tool Failed`

**Comment by facaiy** [comment]
Sanity Checks failed again. Not sure whether the abortion is from Github or build script. Do I need to merge the latest master branch from upstream? Perhaps it could solve the problem.

```bash
[EnvInject] - Loading node environment variables.
Building remotely on cpu33-slave (slave cpu-slave) in workspace /var/lib/jenkins/workspace/tensorflow-pull-requests-sanity
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/tensorflow/tensorflow.git # timeout=10
Fetching upstream changes from https://github.com/tensorflow/tensorflow.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/tensorflow/tensorflow.git +refs/pull/*:refs/remotes/origin/pr/*
Build was aborted
Aborted by unknown
Unable to get pull request builder trigger!!
Setting status of a1faeb5123de7a63e8a432c818092d7b7da56702 to FAILURE with url https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/10748/ and message: 'FAILURE
```

**Comment by facaiy** [comment]
Hi, @jhseu. Could you retest it? Thanks

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
(API review: good to go)

**Comment by facaiy** [comment]
Thanks very much for your review and help, @martinwicke @lukaszkaiser @yifeif .

**Pull request merged**